File: Julia/Gemini2.5/NBodySimulationPar.jl
using LinearAlgebra
using Printf
using Random
using Base.Threads # Added for multi-threading

# --- Constants ---
const G = 1.0             # Gravitational constant (set to 1 for simplicity)
const SOFTENING = 0.01    # Softening parameter to avoid divergences at r=0
const SOFTENING_SQ = SOFTENING^2 # Pre-calculate squared softening

# --- Data Structures ---
# positions: N x 3 matrix (x, y, z for each body)
# velocities: N x 3 matrix (vx, vy, vz for each body)
# masses: N x 1 vector (mass of each body)
# accelerations: N x 3 matrix (ax, ay, az for each body)

# --- Core Physics Calculation (Multithreaded) ---

"""
Calculates the gravitational accelerations for all bodies using multiple threads.
The outer loop over bodies 'i' is parallelized.
"""
function calculate_accelerations!(accelerations::Matrix{Float64},
                                positions::Matrix{Float64},
                                masses::Vector{Float64})
    N = size(positions, 1)
    fill!(accelerations, 0.0) # Reset accelerations (safe before @threads)

    # Parallelize the loop over bodies 'i'
    Threads.@threads for i in 1:N
        # Accumulators for the current body 'i' (local to this iteration/thread)
        accel_x_i = 0.0
        accel_y_i = 0.0
        accel_z_i = 0.0

        # Inner loop: Calculate influence of all *other* bodies 'j' on 'i'
        # Note: We cannot use the j > i optimization directly here because
        # each thread only computes forces *for* its assigned 'i' values.
        # We also don't update 'j' here to avoid race conditions.
        for j in 1:N
            if i == j
                continue # Skip self-interaction
            end

            # Vector from body i to body j
            dx = positions[j, 1] - positions[i, 1]
            dy = positions[j, 2] - positions[i, 2]
            dz = positions[j, 3] - positions[i, 3]

            # Squared distance with softening
            dist_sq = dx^2 + dy^2 + dz^2 + SOFTENING_SQ

            # Inverse cube distance factor (including G * m_j)
            # a_i = sum_{j!=i} G * m_j * (r_j - r_i) / |r_j - r_i|^3 (softened)
            inv_dist_cubed = G / (dist_sq * sqrt(dist_sq)) # Faster than G / dist_sq^(3/2)
            force_factor = masses[j] * inv_dist_cubed

            # Accumulate acceleration on body i due to body j
            accel_x_i += force_factor * dx
            accel_y_i += force_factor * dy
            accel_z_i += force_factor * dz
        end
        # Assign the calculated acceleration to the main array (safe, only this thread writes to index i)
        accelerations[i, 1] = accel_x_i
        accelerations[i, 2] = accel_y_i
        accelerations[i, 3] = accel_z_i
    end
end

# --- Simulation Step ---
# (No changes needed here, calls the threaded acceleration function)
"""
Performs one step of the simulation using the first-order Kick-Step
(Euler-Cromer or semi-implicit Euler) method.
"""
function kick_step_update!(positions::Matrix{Float64},
                          velocities::Matrix{Float64},
                          masses::Vector{Float64},
                          accelerations::Matrix{Float64},
                          dt::Float64)

    # Calculate accelerations based on current positions (uses the threaded version)
    calculate_accelerations!(accelerations, positions, masses)

    # 1. Kick: Update velocities (v_new = v_old + a * dt)
    # Broadcasting is generally efficient, threading might not add much here.
    velocities .+= accelerations .* dt

    # 2. Step: Update positions (p_new = p_old + v_new * dt)
    positions .+= velocities .* dt
end


# --- Energy Calculation (Potential Energy Multithreaded) ---

"""
Calculates the total energy (Kinetic + Potential) of the system.
Potential energy calculation is O(N^2) and parallelized using thread-local storage.
Kinetic energy (O(N)) is calculated sequentially.
"""
function calculate_total_energy(positions::Matrix{Float64},
                                velocities::Matrix{Float64},
                                masses::Vector{Float64})::Float64
    N = size(positions, 1)
    kinetic_energy = 0.0
    potential_energy = 0.0

    # Kinetic Energy: KE = sum(0.5 * m_i * |v_i|^2) - O(N)
    # Calculated sequentially for simplicity, parallelize if it becomes a bottleneck
    for i in 1:N
        vel_sq = velocities[i, 1]^2 + velocities[i, 2]^2 + velocities[i, 3]^2
        kinetic_energy += 0.5 * masses[i] * vel_sq
    end

    # Potential Energy: PE = sum_{i < j} (-G * m_i * m_j / |r_i - r_j|) - O(N^2)
    # Parallelized using per-thread accumulators
    num_threads = Threads.nthreads()
    pe_partials = zeros(Float64, num_threads) # Storage for each thread's partial sum

    Threads.@threads for i in 1:N
        tid = Threads.threadid()
        pe_thread = 0.0 # Accumulator local to this thread's work on this 'i' loop
        # Loop only j > i to avoid double counting and self-interaction
        for j in (i+1):N
            dx = positions[j, 1] - positions[i, 1]
            dy = positions[j, 2] - positions[i, 2]
            dz = positions[j, 3] - positions[i, 3]

            dist_sq = dx^2 + dy^2 + dz^2 + SOFTENING_SQ
            dist = sqrt(dist_sq)

            pe_thread -= G * masses[i] * masses[j] / dist
        end
        # Accumulate the result for this 'i' into the thread's partial sum storage
        # Note: This+= operation on pe_partials[tid] is NOT atomic, but it's safe
        # because each thread only writes to its own index 'tid'.
        # However, multiple 'i' iterations run by the same thread will accumulate correctly.
        # A cleaner way might be to sum pe_thread within the thread and add once at the end,
        # but this accumulation should also work correctly with @threads partitioning.
        # Let's use an atomic add just to be explicitly safe if multiple iterations for
        # the same thread were interleaved in some unexpected scheduler way, though unlikely
        # for simple loop partitioning. Or stick to the simpler += as race conditions
        # between iterations assigned to the *same thread* are not the primary concern.
        # Sticking with += as it's likely correct and avoids atomic overhead.
        pe_partials[tid] += pe_thread
    end

    # Sum the partial potential energies from all threads
    potential_energy = sum(pe_partials)

    return kinetic_energy + potential_energy
end


# --- Initialization ---
# (No changes needed here)
"""
Initializes a system with one central body and N_orbiting smaller bodies
in approximately circular orbits within a specified radius range.
"""
function initialize_circular_orbits(N_orbiting::Int,
                                    central_mass::Float64;
                                    min_radius::Float64 = 1.0,
                                    max_radius::Float64 = 10.0,
                                    orbiting_mass_factor::Float64 = 1e-7 # Mass relative to central
                                    )::Tuple{Matrix{Float64}, Matrix{Float64}, Vector{Float64}}
    N_total = N_orbiting + 1
    positions = zeros(Float64, N_total, 3)
    velocities = zeros(Float64, N_total, 3)
    masses = zeros(Float64, N_total)

    # Initialize central body (at index 1)
    masses[1] = central_mass

    # Initialize orbiting bodies
    rng = MersenneTwister() # Use default global RNG or create one per thread if needed elsewhere
    for i in 2:N_total
        masses[i] = central_mass * orbiting_mass_factor * (0.5 + rand(rng))
        radius = min_radius + (max_radius - min_radius) * rand(rng)
        phi = 2.0 * pi * rand(rng)
        cos_theta = 2.0 * rand(rng) - 1.0
        sin_theta = sqrt(max(0.0, 1.0 - cos_theta^2)) # Ensure non-negative under sqrt

        x = radius * sin_theta * cos(phi)
        y = radius * sin_theta * sin(phi)
        z = radius * cos_theta
        positions[i, :] = [x, y, z]

        vel_magnitude = sqrt(G * central_mass / radius)
        pos_vec = positions[i, :]
        if abs(pos_vec[1]) < 1e-9 && abs(pos_vec[2]) < 1e-9
             axis_vec = [1.0, 0.0, 0.0]
        else
             axis_vec = [0.0, 0.0, 1.0]
        end

        # Handle potential zero vector cross product if pos_vec aligns with axis_vec
        vel_dir = cross(pos_vec, axis_vec)
        if norm(vel_dir) < 1e-9 # If vectors were parallel, try another axis
             axis_vec = [0.0, 1.0, 0.0] # Try Y-axis
             vel_dir = cross(pos_vec, axis_vec)
        end

        # Ensure velocity direction is normalized, handle zero norm case if necessary
        norm_vel_dir = norm(vel_dir)
        if norm_vel_dir > 1e-9
            vel_dir_normalized = vel_dir / norm_vel_dir
        else
            # Handle cases where position is exactly at origin or something went wrong
            # Assigning a random orthogonal direction as fallback
            vel_dir_normalized = normalize([rand(rng)-0.5, rand(rng)-0.5, rand(rng)-0.5])
            println("Warning: Could not determine unique velocity direction for body $i. Using random.")
        end


        velocities[i, :] = vel_magnitude .* vel_dir_normalized
    end

    return positions, velocities, masses
end

# --- Main Simulation Function ---

function run_simulation(num_bodies_orbiting::Int, num_steps::Int, dt::Float64)
    # --- Check Threads ---
    num_threads = Threads.nthreads()
    if num_threads == 1
        println("WARNING: Julia started with only 1 thread. Multithreading disabled.")
        println("         Restart Julia with '-t auto' or '-t N' (e.g., julia -t 4 script.jl)")
    end
    println("--- N-Body Simulation Start (Threads: $num_threads) ---")
    println("Number of orbiting bodies: ", num_bodies_orbiting)
    println("Total bodies: ", num_bodies_orbiting + 1)
    println("Number of steps: ", num_steps)
    println("Time step (dt): ", dt)
    println("Softening factor: ", SOFTENING)
    println("Gravitational Constant (G): ", G)
    println("---------------------------------")

    # Initialization
    println("Initializing system...")
    central_mass = 1.0e6
    positions, velocities, masses = initialize_circular_orbits(num_bodies_orbiting, central_mass)
    N_total = size(positions, 1)
    # Acceleration buffer allocation remains the same
    accelerations = zeros(Float64, N_total, 3)
    println("Initialization complete.")

    # Energy check before simulation
    println("Calculating initial energy...")
    initial_energy = calculate_total_energy(positions, velocities, masses) # Uses threaded PE calc
    @printf("Initial Total Energy: %.6e\n", initial_energy)
    println("---------------------------------")

    # --- Simulation Loop ---
    println("Starting simulation loop...")
    start_time = time()

    for step in 1:num_steps
        kick_step_update!(positions, velocities, masses, accelerations, dt)

        # Optional: Print progress
        if step % max(1, num_steps ÷ 10) == 0 || step == num_steps # Avoid division by zero if num_steps < 10
           elapsed_time = time() - start_time
           @printf("Step: %d / %d (%.1f%%), Elapsed Time: %.2f s\n",
                   step, num_steps, 100 * step / num_steps, elapsed_time)
        end
    end

    end_time = time()
    total_time = end_time - start_time
    println("Simulation loop finished.")
    println("---------------------------------")

    # Energy check after simulation
    println("Calculating final energy...")
    final_energy = calculate_total_energy(positions, velocities, masses) # Uses threaded PE calc
    @printf("Final Total Energy:   %.6e\n", final_energy)

    # Accuracy check
    energy_diff = abs(final_energy - initial_energy)
    relative_error = if abs(initial_energy) > 1e-12
        energy_diff / abs(initial_energy)
    else
        energy_diff
    end
    @printf("Absolute Energy Difference: %.6e\n", energy_diff)
    @printf("Relative Energy Error: %.6e (%.4f%%)\n", relative_error, relative_error * 100)
    println("---------------------------------")
    @printf("Total Simulation Time: %.3f seconds\n", total_time)
    println("--- N-Body Simulation End ---")

end

# --- Run ---
# Remember to start Julia with threads enabled:
# julia -t auto your_script_name.jl
# or
# julia -t N your_script_name.jl (where N is the number of threads)

# **WARNING:** N=1,000,001 is EXTREMELY computationally expensive even multithreaded.
# The complexity is still O(N^2), threading only divides the constant factor.
# Expect very long runtimes. Start with smaller N for testing.

# const NUM_ORBITING_BODIES = 1_000_000 # Requested, but likely infeasible runtime
const NUM_ORBITING_BODIES = 1000      # More reasonable number for testing
const NUM_STEPS = 1000
const DT = 0.001                      # Time step

# Check if running interactively or as script
if isinteractive() && NUM_ORBITING_BODIES > 5000 # Add threshold for warning
    println("Running simulation with N_orbiting = $(NUM_ORBITING_BODIES) for $(NUM_STEPS) steps using $(Threads.nthreads()) threads.")
    println("WARNING: If N_orbiting is large, this will take a very long time even with threads.")
    println("Press Enter to continue, or Ctrl+C to abort.")
    readline()
elseif !isinteractive() && Threads.nthreads() == 1
     println("WARNING: Running script with only 1 thread. Performance will be limited.")
     println("         Consider restarting Julia with the -t flag for multithreading.")
end

run_simulation(NUM_ORBITING_BODIES, NUM_STEPS, DT)

File: Go/Gemini2.5/NBodySimulationPar.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"runtime" // Added for NumCPU
	"sync"    // Added for WaitGroup
	"time"
)

// --- Constants ---
const (
	G                 = 1.0   // Gravitational constant
	dt                = 0.001 // Time step
	epsilon           = 0.01  // Softening factor to avoid division by zero/large forces
	numSteps          = 100   // Number of simulation steps
	numOrbitingBodies = 10000 // Number of orbiting bodies (WARNING: Very large even for parallel!)
	// numOrbitingBodies = 1000 // More reasonable number for testing
)

// --- Data Structures ---

// Vector3D represents a point or vector in 3D space
type Vector3D struct {
	X, Y, Z float64
}

// Body represents a celestial body with mass, position, and velocity
// Using pointers for Position and Velocity might be slightly more idiomatic
// when modifying them in place frequently, but structs work fine here too.
type Body struct {
	Mass     float64
	Position Vector3D
	Velocity Vector3D
}

// --- Vector Operations ---

// Add returns the sum of two vectors
func (v Vector3D) Add(other Vector3D) Vector3D {
	return Vector3D{v.X + other.X, v.Y + other.Y, v.Z + other.Z}
}

// Subtract returns the difference between two vectors (v - other)
func (v Vector3D) Subtract(other Vector3D) Vector3D {
	return Vector3D{v.X - other.X, v.Y - other.Y, v.Z - other.Z}
}

// Scale returns the vector scaled by a scalar factor
func (v Vector3D) Scale(scalar float64) Vector3D {
	return Vector3D{v.X * scalar, v.Y * scalar, v.Z * scalar}
}

// MagnitudeSquared returns the squared magnitude (length) of the vector
func (v Vector3D) MagnitudeSquared() float64 {
	return v.X*v.X + v.Y*v.Y + v.Z*v.Z
}

// Magnitude returns the magnitude (length) of the vector
func (v Vector3D) Magnitude() float64 {
	return math.Sqrt(v.MagnitudeSquared())
}

// Normalize returns a unit vector in the same direction
func (v Vector3D) Normalize() Vector3D {
	mag := v.Magnitude()
	if mag == 0 {
		return Vector3D{0, 0, 0} // Avoid division by zero
	}
	return v.Scale(1.0 / mag)
}

// --- Simulation Logic ---

// calculateTotalEnergy calculates the total kinetic and potential energy of the system
// (Remains sequential as it's mainly for verification before/after)
func calculateTotalEnergy(bodies []Body) float64 {
	kineticEnergy := 0.0
	potentialEnergy := 0.0

	// Kinetic Energy: Sum(0.5 * m * v^2)
	for i := range bodies {
		kineticEnergy += 0.5 * bodies[i].Mass * bodies[i].Velocity.MagnitudeSquared()
	}

	// Potential Energy: Sum(-G * m_i * m_j / |r_i - r_j|) for i < j
	// This O(N^2) part is only run twice, so parallelizing it isn't critical path
	for i := 0; i < len(bodies); i++ {
		for j := i + 1; j < len(bodies); j++ {
			deltaPos := bodies[i].Position.Subtract(bodies[j].Position)
			distance := deltaPos.Magnitude()
			if distance > 0 { // Avoid self-interaction potential if distance is zero
				potentialEnergy -= G * bodies[i].Mass * bodies[j].Mass / distance
			}
		}
	}

	return kineticEnergy + potentialEnergy
}

// --- Parallel Simulation Step ---

// calculateAccelerationChunk calculates acceleration for a subset of bodies
func calculateAccelerationChunk(bodies []Body, accelerations []Vector3D, start, end int, wg *sync.WaitGroup, gravitationalConstant, softeningFactor float64) {
	defer wg.Done() // Signal completion when function exits
	n := len(bodies)
	softeningSq := softeningFactor * softeningFactor

	for i := start; i < end; i++ {
		acc := Vector3D{0, 0, 0} // Initialize acceleration for body i
		for j := 0; j < n; j++ {
			if i == j {
				continue // Don't interact with self
			}
			deltaPos := bodies[j].Position.Subtract(bodies[i].Position) // Vector from i to j
			distSq := deltaPos.MagnitudeSquared()
			invDistCubed := 1.0 / math.Pow(distSq+softeningSq, 1.5) // Include softening

			// Force magnitude G * m_i * m_j / (r^2 + eps^2)^(3/2) * deltaPos
			// Acc = F/m_i = G * m_j / (r^2 + eps^2)^(3/2) * deltaPos
			accScale := gravitationalConstant * bodies[j].Mass * invDistCubed
			acc = acc.Add(deltaPos.Scale(accScale))
		}
		accelerations[i] = acc // Store calculated acceleration
	}
}

// updateVelocityChunk updates velocity for a subset of bodies
func updateVelocityChunk(bodies []Body, accelerations []Vector3D, start, end int, wg *sync.WaitGroup, timeStep float64) {
	defer wg.Done()
	for i := start; i < end; i++ {
		// Use pointer to modify the slice element directly if using []*Body
		// bodies[i].Velocity = bodies[i].Velocity.Add(accelerations[i].Scale(timeStep))
		// If using []Body, we need to assign back to the slice index
		bodies[i].Velocity = bodies[i].Velocity.Add(accelerations[i].Scale(timeStep))
	}
}

// updatePositionChunk updates position for a subset of bodies
func updatePositionChunk(bodies []Body, start, end int, wg *sync.WaitGroup, timeStep float64) {
	defer wg.Done()
	for i := start; i < end; i++ {
		// Use pointer to modify the slice element directly if using []*Body
		// bodies[i].Position = bodies[i].Position.Add(bodies[i].Velocity.Scale(timeStep))
		// If using []Body, we need to assign back to the slice index
		bodies[i].Position = bodies[i].Position.Add(bodies[i].Velocity.Scale(timeStep))
	}
}

// simulateStepParallel performs one step using multiple goroutines
func simulateStepParallel(bodies []Body, accelerations []Vector3D, numWorkers int, timeStep, gravitationalConstant, softeningFactor float64) {
	n := len(bodies)
	var wg sync.WaitGroup

	// --- 1. Parallel Acceleration Calculation ---
	wg.Add(numWorkers)
	chunkSize := (n + numWorkers - 1) / numWorkers // Ceiling division
	for i := 0; i < numWorkers; i++ {
		start := i * chunkSize
		end := start + chunkSize
		if end > n {
			end = n
		}
		if start >= end { // Handle case where numWorkers > n
			wg.Done() // Decrement counter if worker has no work
			continue
		}
		// Pass necessary variables to the goroutine
		go calculateAccelerationChunk(bodies, accelerations, start, end, &wg, gravitationalConstant, softeningFactor)
	}
	wg.Wait() // Wait for all acceleration calculations to finish

	// --- 2. Parallel Velocity Update (Kick) ---
	wg.Add(numWorkers)
	for i := 0; i < numWorkers; i++ {
		start := i * chunkSize
		end := start + chunkSize
		if end > n {
			end = n
		}
		if start >= end {
			wg.Done()
			continue
		}
		go updateVelocityChunk(bodies, accelerations, start, end, &wg, timeStep)
	}
	wg.Wait() // Wait for all velocity updates to finish

	// --- 3. Parallel Position Update (Step) ---
	wg.Add(numWorkers)
	for i := 0; i < numWorkers; i++ {
		start := i * chunkSize
		end := start + chunkSize
		if end > n {
			end = n
		}
		if start >= end {
			wg.Done()
			continue
		}
		go updatePositionChunk(bodies, start, end, &wg, timeStep)
	}
	wg.Wait() // Wait for all position updates to finish
}

// --- Initialization --- (Same as before)

// initializeCircularOrbitSystem creates a central body and N orbiting bodies
func initializeCircularOrbitSystem(numOrbiting int, centralMass, orbitingMass, minRadius, maxRadius float64) []Body {
	bodies := make([]Body, numOrbiting+1)
	// Use a fixed seed for reproducibility during testing if needed
	// rand.Seed(12345)
	rand.Seed(time.Now().UnixNano()) // Seed random number generator

	// Central Body (at origin, initially stationary)
	bodies[0] = Body{
		Mass:     centralMass,
		Position: Vector3D{0, 0, 0},
		Velocity: Vector3D{0, 0, 0},
	}

	// Orbiting Bodies
	for i := 1; i <= numOrbiting; i++ {
		radius := minRadius + rand.Float64()*(maxRadius-minRadius)
		theta := 2 * math.Pi * rand.Float64()
		phi := math.Acos(1 - 2*rand.Float64())
		posX := radius * math.Sin(phi) * math.Cos(theta)
		posY := radius * math.Sin(phi) * math.Sin(theta)
		posZ := radius * math.Cos(phi)
		positionVec := Vector3D{posX, posY, posZ}

		orbitSpeed := math.Sqrt(G * centralMass / radius)

		// Simplified perpendicular velocity calculation (assuming Z is up)
		// For a position vector (x, y, z), a perpendicular vector in the XY plane projection's
		// perpendicular direction is (-y, x, 0). Normalize and scale.
		// This prefers orbits somewhat aligned with the XY plane initially.
		// A more robust method (like the cross-product one) ensures better 3D distribution.
		// Let's stick to the cross-product method for better generality:
		helperVec := Vector3D{rand.Float64()*2 - 1, rand.Float64()*2 - 1, rand.Float64()*2 - 1}.Normalize()
		// Avoid parallel vectors (cross product would be zero)
		dotProd := positionVec.X*helperVec.X + positionVec.Y*helperVec.Y + positionVec.Z*helperVec.Z
		if math.Abs(dotProd) > 0.99 { // If vectors are nearly parallel/anti-parallel
			helperVec = Vector3D{0, 1, 0}       // Use a default perpendicular seed vector
			if math.Abs(positionVec.Y) > 0.99 { // If position is along Y axis, use X axis
				helperVec = Vector3D{1, 0, 0}
			}
		}

		velDirection := Vector3D{
			positionVec.Y*helperVec.Z - positionVec.Z*helperVec.Y,
			positionVec.Z*helperVec.X - positionVec.X*helperVec.Z,
			positionVec.X*helperVec.Y - positionVec.Y*helperVec.X,
		}.Normalize()

		velocityVec := velDirection.Scale(orbitSpeed)

		bodies[i] = Body{
			Mass:     orbitingMass,
			Position: positionVec,
			Velocity: velocityVec,
		}
	}

	return bodies
}

// --- Main Function ---

func main() {
	numWorkers := runtime.NumCPU()
	runtime.GOMAXPROCS(numWorkers) // Ensure Go uses all available cores

	fmt.Printf("Starting N-body simulation (Parallel)...\n")
	fmt.Printf("Using %d CPU cores/workers.\n", numWorkers)
	fmt.Printf("Number of bodies: %d (+1 central body)\n", numOrbitingBodies)
	fmt.Printf("Number of steps: %d\n", numSteps)
	fmt.Printf("Time step (dt): %.4f\n", dt)
	fmt.Printf("WARNING: N=%d is very large. Even parallelized, this O(N^2) simulation may take an extremely long time.\n", numOrbitingBodies+1)

	// --- Initialization ---
	startTime := time.Now()
	centralMass := 10000.0 // Mass of the central body
	orbitingMass := 0.01   // Mass of the smaller orbiting bodies
	minRadius := 5.0
	maxRadius := 50.0

	bodies := initializeCircularOrbitSystem(numOrbitingBodies, centralMass, orbitingMass, minRadius, maxRadius)
	initializationTime := time.Since(startTime)
	fmt.Printf("System initialized in %v\n", initializationTime)

	// Pre-allocate acceleration slice (avoids allocation in loop)
	accelerations := make([]Vector3D, len(bodies))

	// --- Initial Energy Calculation ---
	initialEnergy := calculateTotalEnergy(bodies)
	fmt.Printf("Initial Total Energy: %.5e\n", initialEnergy)

	// --- Simulation Loop ---
	fmt.Println("Running simulation...")
	simulationStartTime := time.Now()

	for step := 0; step < numSteps; step++ {
		// Pass the pre-allocated accelerations slice
		simulateStepParallel(bodies, accelerations, numWorkers, dt, G, epsilon)
		if (step+1)%10 == 0 || step == 0 { // Print progress more often initially/periodically
			fmt.Printf("Step %d/%d completed (Elapsed: %v)\n", step+1, numSteps, time.Since(simulationStartTime))
		}
	}

	simulationTime := time.Since(simulationStartTime)
	fmt.Printf("Simulation completed in %v\n", simulationTime)

	// --- Final Energy Calculation ---
	finalEnergy := calculateTotalEnergy(bodies)
	fmt.Printf("Final Total Energy:   %.5e\n", finalEnergy)

	// --- Energy Verification ---
	energyChange := finalEnergy - initialEnergy
	relativeEnergyChange := math.Abs(energyChange / initialEnergy)
	fmt.Printf("Absolute Energy Change: %.5e\n", energyChange)
	fmt.Printf("Relative Energy Change: %.5e (%.4f%%)\n", relativeEnergyChange, relativeEnergyChange*100.0)

	totalTime := time.Since(startTime)
	fmt.Printf("Total execution time: %v\n", totalTime)

}
